### PR TITLE
VPN-3423 Verify subscription before connecting

### PR DESCRIPTION
**Pull request by Gela**
## Description

This enhances the user experience when a subscription expires.

- Introduce a new state `StateCheckSubscription` which creates a network request to check user account data and make sure their subscription is still valid.
- Add a functional test which overrides the Guardian endpoint to mark a subscription as inactive __after__ the client has authenticated but before they toggle the VPN on. 
- Add a new `QmlQueryComposer` called `SUBSCRIPTION_NEEDED` in queries.js to enable testing this view (apparently we don't have any tests currently that test an expired subscription??)

## Reference

    https://mozilla-hub.atlassian.net/browse/VPN-3423

## Checklist
    
- [x] My code follows the style guidelines for this project
- [x] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [x] I have performed a self review of my own code
- [x] I have commented my code PARTICULARLY in hard to understand areas
- [x] I have added thorough tests where needed

### CODE REVIEW

File: src/apps/vpn/controller.cpp [hunk 1]
```diff
@@ -733,7 +759,7 @@ void Controller::statusUpdated(const QString& serverIpv4Gateway,
 
   list.swap(m_getStatusCallbacks);
   for (const std::function<void(
-           const QString&serverIpv4Gateway, const QString&deviceIpv4Address,
+           const QString& serverIpv4Gateway, const QString& deviceIpv4Address,
```
Comment (Gela) [inline src/apps/vpn/controller.cpp:762]: Automatic (and unrelated) change due to the Format hook

File: tests/functional/testSubscription.js [hunk 2]
```diff
@@ -766,7 +843,7 @@ describe('Subscription view', function() {
               queries.screenSettings.subscriptionView.CANCELLED_LABEL.visible(),
               'text') === data.subscription.expected.label);
 
-      if (data.subscription.value._subscription_type == "web") {
+      if (data.subscription.value._subscription_type == 'web') {
```
Comment (Gela) [inline tests/functional/testSubscription.js:846]: Automatic (and unrelated) change due to the Format hook

File: tests/functional/testSubscription.js [hunk 3]
```diff
@@ -1139,73 +1216,72 @@ describe('Subscription view', function() {
...
-          await vpn.flipFeatureOn('bundleUpgrade');
-        }
-
-        await clickSettingsIcon();
-        await openSubscriptionManagement();
-
-        await vpn.waitForQuery(queries.screenSettings.subscriptionView
-                                   .RELAY_UPSELL_STATUS.hidden());
-      });
+     async () => {
```
Comment (Gela) [inline tests/functional/testSubscription.js:1219]: Automatic (and unrelated) change due to the Format hook

File: src/apps/vpn/controller.h [hunk 4]
```diff
@@ -186,6 +188,8 @@ class Controller final : public QObject {
   void serverDataChanged();
 
  private:
+  ErrorHandler::ErrorPropagationPolicy m_errorPropagationPolicy =
```
Comment (birdsarah) [inline src/apps/vpn/controller.h:191]: I don't think this is needed anymore.

File: src/apps/vpn/controller.cpp [hunk 5]
```diff
@@ -219,6 +220,30 @@ bool Controller::activate(const ServerData& serverData,
...
+    // replicate the behavior of a TaskAccount.
+    TaskFunction* task = new TaskFunction([]() {});
+    NetworkRequest* request = new NetworkRequest(task, 200);
+    request->auth(MozillaVPN::authorizationHeader());
+    request->get(AppConstants::apiUrl(AppConstants::Account));
+
+    connect(request, &NetworkRequest::requestFailed, this,
+            [task, this](QNetworkReply::NetworkError error, const QByteArray&) {
+              logger.error() << "Account request failed" << error;
+              REPORTNETWORKERROR(error, m_errorPropagationPolicy, task->name());
```
Comment (birdsarah) [inline src/apps/vpn/controller.cpp:239]: I'm thinking this should be something like

```
REPORTNETWORKERROR(error, ErrorHandler::DoNotPropagateError, "PreActivationSubscriptionCheck");
```

File: tests/functional/testSubscription.js [hunk 6]
```diff
@@ -26,6 +27,82 @@ const SUBSCRIPTION_DETAILS = {
...
+
+    this.ctx.guardianOverrideEndpoints = {
+      GETs: {
+        '/api/v1/vpn/account': {
+          status: 200,
+          requiredHeaders: ['Authorization'],
+          body: userDataActive
+        }
+      },
+      POSTs: {
```
Comment (birdsarah) [inline tests/functional/testSubscription.js:68]: I think it's worth having a comment here about how the device data needs to match in all these cases.

File: tests/functional/testSubscription.js [hunk 7]
```diff
@@ -26,6 +27,82 @@ const SUBSCRIPTION_DETAILS = {
...
+      this.ctx.guardianOverrideEndpoints.GETs['/api/v1/vpn/account'].body =
+          userDataInactive;
+
+      // Step 3: Attempt to toggle the VPN on.
+      await vpn.waitForQuery(queries.screenHome.CONTROLLER_TITLE.visible());
+      await vpn.clickOnQuery(queries.screenHome.CONTROLLER_TOGGLE.visible());
+
+      // Verify that user gets the "Subscribe to Mozilla VPN" screen.
+      await vpn.waitForQuery(queries.screenHome.SUBSCRIPTION_NEEDED.visible());
+    });
```
Comment (birdsarah) [inline tests/functional/testSubscription.js:102]: Please add a test that covers the case where the call to `/api/v1/vpn/account` fails. I think my expectation is that it does not affect the VPN continuing with activation.

File: tests/functional/testSubscription.js [hunk 8]
```diff
@@ -26,6 +27,82 @@ const SUBSCRIPTION_DETAILS = {
...
+      // "Subscribe to Mozilla VPN" screen.
+
+      // Step 1: User logs into the client successfully.
+      await vpn.authenticateInApp(true, true);
+
+      // Step 2: Override the Guardian endpoint to mock an expired subscription.
+      this.ctx.guardianOverrideEndpoints.GETs['/api/v1/vpn/account'].body =
+          userDataInactive;
+
+      // Step 3: Attempt to toggle the VPN on.
```
Comment (birdsarah) [inline tests/functional/testSubscription.js:96]: Let's use the vpn.activate() helper method unless you had a reason for not.

File: tests/functional/testSubscription.js [hunk 9]
```diff
@@ -26,6 +27,82 @@ const SUBSCRIPTION_DETAILS = {
...
+      }
+    };
+
+    it('Verify subscription before enabling VPN', async () => {
+      // This test verifies the case where a user is logged in
+      // but the VPN is off when their subscription expires.
+      // When they try to turn the VPN on, they get the
+      // "Subscribe to Mozilla VPN" screen.
+
+      // Step 1: User logs into the client successfully.
```
Comment (birdsarah) [inline tests/functional/testSubscription.js:89]: I love how beautifully commented this test is. 

However, the methods authenticateInApp and activate are very self explanatory. So I would only keep the Step 2 and 4 comments.

**Review by birdsarah — CHANGES_REQUESTED** [review]

File: tests/functional/queries.js [hunk 10]
```diff
@@ -176,7 +180,8 @@ const screenAuthenticationInApp = {
       new QmlQueryComposer('//authVerificationSessionByTotpNeeded-getHelpLink'),
   AUTH_TOTP_BACK_BUTTON:
       new QmlQueryComposer('//authVerificationSessionByTotpNeeded-backButton'),
-  AUTH_TOTP_PASTE_BUTTON: new QmlQueryComposer('//authVerificationSessionByTotpNeeded-inputPasteButton'),
+  AUTH_TOTP_PASTE_BUTTON: new QmlQueryComposer(
```
Comment (bakulf) [inline tests/functional/queries.js:183]: Why this changes? We should use clang-format

**Review by bakulf — CHANGES_REQUESTED** [review]
Well done. Consider deleting the task obj using the "destroyed" signal of the network request. And maybe move that block of code in a separate method.

File: tests/functional/testSubscription.js (hunk 8 above)
Comment (Gela) [reply tests/functional/testSubscription.js:96]: `vpn.activate()` implementation explicitly checks that the VPN was enabled successfully (which doesn't/shouldn't happen in this case) so it ends up error-ing out if I use it here. Instead I need to explicitly attempt to turn the VPN on, then check that we land on the "Subscription required" screen.

File: tests/functional/queries.js (hunk 10 above)
Comment (Gela) [reply tests/functional/queries.js:183]: This is the result of running clang format on this file. I think it moved `//authVerificationSessionByTotpNeeded...` to a new line since it exceeds the line length

File: src/apps/vpn/controller.cpp (hunk 5 above)
Comment (Gela) [reply src/apps/vpn/controller.cpp:239]: I think we should go back to `StateOff` but let me know if you disagree

File: src/apps/vpn/controller.cpp [hunk 11]
```diff
@@ -219,7 +220,38 @@ bool Controller::activate(const ServerData& serverData,
...
+    NetworkRequest* request = new NetworkRequest(task, 200);
+    request->auth(MozillaVPN::authorizationHeader());
+    request->get(AppConstants::apiUrl(AppConstants::Account));
+
+    connect(request, &NetworkRequest::requestFailed, this,
+            [this](QNetworkReply::NetworkError error, const QByteArray&) {
+              logger.error() << "Account request failed" << error;
+              REPORTNETWORKERROR(error, ErrorHandler::DoNotPropagateError,
+                                 "PreActivationSubscriptionCheck");
+              if (MozillaVPN::instance()->state() == MozillaVPN::StateMain) {
```
Comment (bakulf) [inline src/apps/vpn/controller.cpp:241]: Write a comment saying that, we need to see if the error propagation has changed the Mozilla VPN state. We continue only if the user is still authenticated and subscribed.

File: tests/functional/queries.js [hunk 12]
```diff
@@ -143,7 +144,8 @@ const screenAuthenticationInApp = {
   AUTH_SIGNIN_BUTTON: new QmlQueryComposer('//authSignIn-button'),
   AUTH_SIGNIN_GET_HELP_LINK: new QmlQueryComposer('//authSignIn-getHelpLink'),
   AUTH_SIGNIN_BACK_BUTTON: new QmlQueryComposer('//authSignIn-backButton'),
-  AUTH_SIGNIN_PASSWORD_PASTE_BUTTON: new QmlQueryComposer('//authSignIn-inputPasteButton'),
+  AUTH_SIGNIN_PASSWORD_PASTE_BUTTON:
```
Comment (bakulf) [inline tests/functional/queries.js:147]: Are these changes from clang-format?

**Review by bakulf — APPROVED** [review]

File: tests/functional/queries.js (hunk 12 above)
Comment (Gela) [reply tests/functional/queries.js:147]: They are!

File: src/apps/vpn/controller.h [hunk 13]
```diff
@@ -13,6 +13,7 @@
 #include <QTimer>
 #include <functional>
 
+#include "errorhandler.h"
```
Comment (birdsarah) [inline src/apps/vpn/controller.h:16]: Should this import be in the .cpp?

File: tests/functional/testSubscription.js [hunk 14]
```diff
@@ -26,6 +27,114 @@ const SUBSCRIPTION_DETAILS = {
...
+      // When they try to turn the VPN on, they get the
+      // "Subscribe to Mozilla VPN" screen.
+
+      await vpn.authenticateInApp(true, true);
+
+      // Step 1: Override the Guardian endpoint to mock an expired subscription.
+      this.ctx.guardianOverrideEndpoints.GETs['/api/v1/vpn/account'].body =
+          userDataInactive;
+
+      // Step 2: Attempt to toggle the VPN on. This should fail.
```
Comment (birdsarah) [inline tests/functional/testSubscription.js:102]: "This should fail" seems confusing. The next step contains the expectation.

File: tests/functional/testSubscription.js [hunk 15]
```diff
@@ -26,6 +27,114 @@ const SUBSCRIPTION_DETAILS = {
...
+      await vpn.waitForQuery(queries.screenHome.CONTROLLER_TITLE.visible());
+      await vpn.clickOnQuery(queries.screenHome.CONTROLLER_TOGGLE.visible());
+
+      // Step 3: Verify that user gets the "Subscribe to Mozilla VPN" screen.
+      await vpn.waitForQuery(queries.screenHome.SUBSCRIPTION_NEEDED.visible());
+    });
+
+    it('Enable VPN with 500 error', async () => {
```
Comment (birdsarah) [inline tests/functional/testSubscription.js:110]:
```suggestion
    it('Continues to try connecting if call to check subscription status fails', async () => {
```

File: tests/functional/testSubscription.js [hunk 16]
```diff
@@ -26,6 +27,114 @@ const SUBSCRIPTION_DETAILS = {
...
+            device.name = req.body.name;
+            device.pubkey = req.body.pubkey;
+            device.unique_id = req.body.unique_id;
+          },
+          body: {}
+        },
+      }
+    };
+
+    it('Verify subscription before enabling VPN', async () => {
```
Comment (birdsarah) [inline tests/functional/testSubscription.js:90]:
```suggestion
    it('Prompts SubscriptionNeeded if a user's subscription is inactive on connection', async () => {
```
The test framework .describe( .it() ) are trying to nudge you to describe the expected behavior in the test name.

File: tests/functional/testSubscription.js [hunk 17]
```diff
@@ -26,6 +27,114 @@ const SUBSCRIPTION_DETAILS = {
...
+
+      // Step 1: Override the Guardian endpoint to mock an expired subscription.
+      // Set the error status to 500.
+      this.ctx.guardianOverrideEndpoints.GETs['/api/v1/vpn/account'].body =
+          userDataInactive;
+      this.ctx.guardianOverrideEndpoints.GETs['/api/v1/vpn/account'].status =
+          500;
+
+      await vpn.activate();
+      await vpn.waitForCondition(async () => {
```
Comment (birdsarah) [inline tests/functional/testSubscription.js:130]: I would prefer if this waitForCondition was for the `Connecting ...` text not the `VPN is on.` It's not expected that the VPN would successfully connect in this state.

**Review by birdsarah — APPROVED** [review]
Approve with some small changes requested.

File: tests/functional/testSubscription.js (hunk 17 above)
Comment (Gela) [reply tests/functional/testSubscription.js:130]: From my conversation with @bakulf this morning, I believe the VPN should successfully connect in this case. Baku, can you confirm?
